ddtrace/internal/processor/trace.py

```diff
@@ -67,6 +68,27 @@ def process_trace(self, trace):
         return trace
 
 
+@attr.s
+class TraceTagsProcessor(TraceProcessor):
+    """Processor that applies trace-level tags to the trace."""
+
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        if not trace:
+            return trace
+
+        chunk_root = trace[0]
+        ctx = chunk_root._context
+        if not ctx:
+            return trace
+
+        if ctx.dd_origin is not None:
+            for span in trace:
+                span.meta[ORIGIN_KEY] = ctx.dd_origin
```

Comment:
PR looks good. Just a question regarding setting the tag on each span: do we plan to try and do this from Cython with a follow-up change? I guess it makes more sense to have this here, since this tagging is probably not encoding-specific, but there might be some performance gains, considering that benchmark figures seem to indicate a noticeable performance impact?

Reply:
Do you mean implementing this via the MsgpackEncoder / Packer at the Cython level? I agree with you that if our encoding was modified to interact with the buffer directly, this would most likely lead to performance gains. But that would require us to make those changes to the encoding beforehand.

Reply:
> considering that benchmark figures seem to indicate a noticeable performance impact?

which specific numbers are you looking at for this? the _dd.origin vs not for 50 tags? the numbers seem... close/margin of error level, but would love you interpretation too!

Reply:
@brettlangdon I was just elaborating on this line from the PR description:

> Benchmark tests show around 5% increased overhead on origin tagging at best case measurements.

As for the numbers I see from the benchmarks, all have very large variance so they are all statistically compatible with each other (all well within 1 σ). With #2335 merged, we could get some profiling data to see what the impact on an actual app looks like, hopefully with less noise.

Reply:
> With #2335 merged, we could get some profiling data to see what the impact on an actual app looks like, hopefully with less noise.

@P403n1x87 is this something you and @Yun-Kim can work on locally together to verify? If the results are more consistent and we have a quick path to validate at least locally, seems like a worthwhile effort to me. What do you think?

Reply:
Yep, I have a local setup that we could use to get that sort of data out.